### ISSUE TITLE
Documentation for HTTP wrapper

### ISSUE THREAD

**Issue by dukedougal** [techfort/LokiJS#272]
I couldn't see the documentation for the HTTP wrapper although it is mentioned in the features list.  Would you mind pointing me in the right direction please?

**Comment by obeliskos** [comment]
Http server and client were on roadmap, that is primarily an interest of @techfort's. Is that what you are referring to?  I am not familiar with the existence of a loki HTTP wrapper.

**Comment by techfort** [comment]
@dukedougal as @obeliskos mentioned this is in the roadmap. There is an incomplete draft at https://github.com/techfort/snaptun which i only parked due to lack of available time but I am actively looking for contributors to get the project moving along. There has been a few requests for this so i might try to make that time available but with a 5 week old baby i can't make any promise :)

**Comment by johanneslumpe** [comment]
@techfort Did you also had in mind to add user auth etc. To the mix? I've not yet used loki actively but it does look promising. I was thinking trying it out for local networking apps and being able to spawn a master app which allows other networking instances to connect to the shared loki db would be nice. Furthermore I would maybe be able to help out with some things.

**Comment by techfort** [comment]
@johanneslumpe yes absolutely. My dream scenario is to create both a tcp and a http interface for LokiJS and include authentication. I'm aiming for the stars here by trying to get Loki as stable and full-featured as possible so the first pure-javascript stack (as far as I know LokiJS would be the first db server written in javascript). Contributions are really welcome as i'm seriously lacking time lately - what with a 3 month old baby and all.

**Comment by johanneslumpe** [comment]
Yeah I read that above - congrats by the way :)

I was thinking about keeping a `dbusers` collection which would only be queryable by users flagged as admin. My ideal scenario would be that people can log into an app and the same credentials can be used to query the db - yielding only access to collections which they have been granted access too (think about a single server which has multiple namespaced sets of the same collections). If everything uses https, then this might be be fine and there would be no need for tokens or anything. But keeping the connection open (tcp) would of course be better. It would also mitigate the username/pass issue and probably increase performance, with the downside of being more low level. Being able to just fire off a `GET` request would be great though, instead of having to manage the direct tcp connection, because we can just use normal status codes when returning results.

**Comment by techfort** [comment]
@johanneslumpe http and tcp both have "pro et contra", hence why i'm developing (or at least i was developing) them separately from Loki and from each other. TCP has turned out to be more painful, but I don't want to pose limitations on the usage one can imagine to do of LokiJS, On the other hand, a simple JSON based protocol transmitted over the wire as buffer works just fine so it's really not so bad. Any ideas are more than welcome!

**Comment by johanneslumpe** [comment]
@techfort True - supporting both would be great. For TCP it might be sensible to have 3 packages, like `loki-message-formatter`, `loki-tcp-server` and `loki-tcp-client`. The first being a dependency of the other two. Sending basic JSON messages works fine (I just tested that locally with a very basic setup). Things like $regex would require us to pre-parse the message for those occurrences and replace them with a `toString() version of the regex and then later instantiate a new regex again on the server.

The protocol could be as simple as `{ col: 'collection', query: { /* ... */ }, method: 'find', qId: 'someQueryId' }` with individual messages terminated by a line-break or a custom sentinel value. Each method could have it's own parsing function which expects data in a certain way. We'd have `parseFindQuery`, `createFindQuery`, `parseByQuery`, `createByQuery`, etc. So the tcp client would basically have the same interface as a loki db instance, where each of the methods constructs the messages under the hood using the formatting methods and ships it off to the server. Those methods could return promises which get resolved when we get the result for our query.

**Comment by techfort** [comment]
hey @johanneslumpe have you taken a look at https://github.com/techfort/lokijs-server ? Pretty much the direction i have taken is not too far away from what you were describing. I would really be interesting in progressing this, if i have someone to give me a hand given the radical lack of spare time in this period. It will get better soon, hopefully.

**Comment by johanneslumpe** [comment]
@techfort Now I have! You are right, we're thinking about a similar solution. I personally would rather have a more "functional" approach, instead of `new`ing stuff. Promises instead of callbacks (automatically async). I'm also a sucker for ES6/7 syntax. Using numbers to express the operations is a lot better than my crude string idea above.

In terms of an api for the tcp client I was thinking about something like this:
```js
lokiClientInstance
  .getCollection('collection')
  .find({ name: 'Hel' })
  .then(data => { /* ... */ })
  .catch(err => { /* ... */ })
```
Internally this would be translated into a query message object and when `find` is executed we query the server and return a promise which will resolve with the result. Optionally we could add a `isRequired` flag to `find`, which would then reject the promise if no dataset is found. I didn't yet dive too much into the internals of loki, so I'm not sure whether things like sorting would be on the client or the server. But my gut tells me that if we do things like `getCollection('users').sort('createdAt').skip(10).limit(5).find()` it has to be sorted on the server.

Have you thought about the implications of multiple connections querying the same server? Event-loop blocking for example when someone queries against a collection with 100k rows and gets an initial result set of 30k and applies a sort over 3 fields.

**Comment by techfort** [comment]
@johanneslumpe 100% agree on a more functional approach. Amazing how things change in one year, i am a complete converted to ES2015 + FP, so yes, I'd go that way. I have not though about multiple connections - i guess i wanted to have a working prototype before i considered the performance implications. Any thoughts on your part?

**Comment by johanneslumpe** [comment]
@techfort One idea might be to split up a query into a "pipeline", where each of the parts is performed in a separate tick, allowing other actions to run in between. Or the server could automatically spawn a configurable number of workers which deal with the blocking things like sorting etc. Not sure how smooth it would be to push the data to those workers though because it would need to be serialized, which itself is a rather slow task.

How does loki handle sorting? If we take the above query as an example, where the `users` collection has 10k rows. If the `createdAt` field isn't indexed, loki has to fetch all rows, sort them in memory and then return result 11-15 right? If the field is indexed and the index is already sorted then the operation would obviously be a lot faster. Is this about right?

Another thing: what about mapReduce operations? If we allow functions to be serialized and then `eval`d on the server we will have to make sure that they aren't exploiting the system. Some kind of parser/validator (filter out `this`/`window`/`global` etc., basically making sure that a function only uses the passed arguments)

**Comment by techfort** [comment]
@johanneslumpe i think the first solution you proposed looks very attractive. i'm also toying with the idea of saving huge sorted datasets to disk and utilizing I/O to "unblock" the process,, but it's still not clear in my head how to achieve this.

Sorting operations are by default placed after filters but before offset-limit conditions so to enable lazy fetching but @obeliskos can expand on this a little more. 

Maybe before getting bogged into details of the implementation we could rewrite the wrapper in a more ES2015 + FP fashion, and start playing with multiple connections, replication etc. - and see how it goes? i'm all ears of course.

**Comment by johanneslumpe** [comment]
@techfort That makes sense yeah. Getting a solid base done which works well and upon which we can build. I will toy around with some ideas locally a little bit during this week and can post back here as soon as I pushed a first version, how does that sound? Just to get some ideas written out so we can discuss more about the direction using some actual code.

**Comment by techfort** [comment]
@johanneslumpe cool - do you want contributor rights on snaptun and lokijs-server? i'd be happy to do so.

**Comment by johanneslumpe** [comment]
@techfort let me first dabble a bit on my own repo and then if you like what you see we can merge/transfer stuff, ok?

**Comment by techfort** [comment]
@johanneslumpe sure - sounds good!

**Comment by johanneslumpe** [comment]
@techfort started to play around a bit: https://github.com/johanneslumpe/lokijs-tcp-server tell me what you think about that. It's in no way near perfect, but shows some direction.

**Comment by techfort** [comment]
@johanneslumpe this is awesome stuff, I love the direction you took! This is definitely promising stuff and i'd be more than happy to see this evolving into a fully fledged product. @obeliskos can you take a look too? we had an exchange last year about this so i'd like to hear what you have to say (i'm aware it's probably not your use case, but still..)

**Comment by johanneslumpe** [comment]
@techfort thanks, appreciate it! Let's see what @obeliskos will say :+1:

**Comment by obeliskos** [comment]
Pretty cool stuff, nice!  Feel free to pursue this however you (two) feel most comfortable with.  I can understand your pipeline reasoning as that provides you a level of scheduling in an otherwise synchronous query.  For different reasons your pipeline is very similar to collection transforms.   

ES6 syntax and promise structures make for clean logic, and for a server I should think you could add any dependencies you want, so looks great to me.

I have only considered experimenting with loki along with express services but that would be mainly for personal use/learning.

Look forward to seeing what you guys come up with, if you need to ask me anything feel free :)  Otherwise I will monitor and chime in if I see anything.

**Comment by techfort** [comment]
hey @johanneslumpe - the tcp server implementation i had there was not even a draft, i'm happy to keep it for archive purpose but it would probably be a good idea to start a new repo from scratch. We need a codename (like the http wrapper for lokijs is called Snaptun) - any ideas? @obeliskos ?

**Comment by johanneslumpe** [comment]
@techfort Do your codenames follow a specific pattern?

**Comment by obeliskos** [comment]
Generally names related to Norse mythology or Loki himself.  Any ideas? If you need suggestions I have a few :)

**Comment by johanneslumpe** [comment]
@obeliskos I'm open for ideas :) I know some Norse mythology but I'm not really well versed in it. The pipeline could be Sleipnir, because of its eight legs maybe.

**Comment by techfort** [comment]
@johanneslumpe @obeliskos Sleipnir sounds good - as it's one of Loki's children :)

**Comment by johanneslumpe** [comment]
@techfort cool :) Then the pipeline package shall be named Sleipnir! We need a name for the tcp server itself too @obeliskos

**Comment by obeliskos** [comment]
Ah ok, well whoever works on it the most should get final call.  I was just going to recommend some general names like Bifrost (or heimdal) or maybe less categorical 'huginn or muninn' (Odin's ravens)... but spleipnir is good.  Or depending on how you break up your modules you could apply any of those.

**Comment by techfort** [comment]
@obeliskos I'm keeping muninn for the next in memory db, because as you
surely know, muninn means "memory" :-D

**Comment by johanneslumpe** [comment]
@Techfort next as in loki 2.0 or a completely different thing?

**Comment by techfort** [comment]
@johanneslumpe a completely different thing i think. Wouldn't want to piss mighty Loki off by renaming this one! I'm writing a nosql db for android, but i already named that sigyn, so who knows.

**Comment by johanneslumpe** [comment]
@techfort So basically Loki is already EOL?

**Comment by techfort** [comment]
@johanneslumpe not at all! when i talk about other in-memory dbs i am talking anything from a java db for android to a c++ stream processor - LokiJS will outlive the Ragnarok. Sorry for the confusion there :)

**Comment by obeliskos** [comment]
There can be only one loki.js!.. wait that's wrong mythology.  I think your safe from worrying about that for awhile.   Norse mythology on the other hand seems always in flux.

**Comment by johanneslumpe** [comment]
@techfort ah ok, wasn't sure about that - thanks for the update :) So should the whole tcp server be Sleipnir for now? And then later we refactor the pure pipeline part out into a separate repo, to be used with the http server?

**Comment by techfort** [comment]
@obeliskos @johanneslumpe yes let's name sleipnir the tcp server. I created
the repo already, i'll edit the description.

**Comment by johanneslumpe** [comment]
@techfort sorry for not yet committing anything - have some stuff to deal with right now. Will do it as soon as possible.

**Comment by techfort** [comment]
hey - no rush.
